auth/oidc.go

```diff
@@ -0,0 +1,218 @@
+)
+
+// OIDC represents our authentication using openid connect and required
+// dependencies like logger.
+type OIDC struct {
+	oidcProvider
+	clientID string
+	logger   log.Logger
+}
+
+// oidcProvider required interface of oidc dependecy.
+type oidcProvider interface {
+	Verifier(config *oidc.Config) *oidc.IDTokenVerifier
+	Endpoint() oauth2.Endpoint
+}
+
+// New returns a new OIDC using the provided clientID and issuer to validate
+// incoming authentication tokens at OIDC.Auth http middleware method.
+func New(log log.Logger, clientID, issuer string) (*OIDC, error) {
+	p, err := oidc.NewProvider(context.Background(), issuer)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create oidc provider")
+	}
+
+	return &OIDC{
+		oidcProvider: p,
+		logger:       log,
+		clientID:     clientID,
+	}, nil
+}
+
+// Auth is a middleware used to validate authorization token and populate context
+// with stardard claims.
+func (o *OIDC) Auth(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		rawAccessToken := r.Header.Get("Authorization")
+		if rawAccessToken == "" {
+			http.Error(w, "no authorization header", http.StatusBadRequest)
+			o.logger.Warn(r.Context(), "auth: empty authorization handler")
+
+			return
+		}
+
+		parts := strings.Split(rawAccessToken, " ")
+		if len(parts) != tokenParts {
+			w.WriteHeader(http.StatusBadRequest)
+			o.logger.Warn(r.Context(), "auth: unexpected  authorization size")
+
+			return
+		}
+
+		idToken, err := o.Verifier(&oidc.Config{
+			ClientID:             o.clientID,
+			Now:                  time.Now,
+			SupportedSigningAlgs: nil,
+			SkipClientIDCheck:    false,
+			SkipExpiryCheck:      false,
+			SkipIssuerCheck:      false,
+		}).Verify(r.Context(), parts[1])
+		if err != nil {
+			http.Error(w, "cannot validate token: "+err.Error(), http.StatusUnauthorized)
+			o.logger.Warn(r.Context(), "auth: invalid token", log.Any("token", parts[1]), log.Error(err))
+
+			return
+		}
+
+		resp := Claims{}
+		if err := idToken.Claims(&resp); err != nil {
+			http.Error(w, "cannot unmarshal claims: "+err.Error(), http.StatusInternalServerError)
+			o.logger.Warn(r.Context(), "auth: cannot unmarshal claims", log.Error(err))
+
+			return
+		}
+
+		ctx := context.WithValue(r.Context(), contextKey, resp)
+
+		next.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+// HasRole inspect ctx for role claim, if present returns true. Before using this
+// function ctx should be populated with the token claims, generally using
+// OIDC.Auth method.
+func HasRole(ctx context.Context, role string) bool {
+	r := ctx.Value(contextKey)
+	if r == nil {
+		return false
+	}
+
+	claims, ok := r.(Claims)
+	if !ok {
+		return false
+	}
+	access, ok := claims["realm_access"].(map[string]interface{})
+	if !ok {
+		return false
+	}
+	roles, ok := access["roles"].([]interface{})
+	if !ok {
+		return false
+	}
+
+	found := false
+	for _, r := range roles {
+		if role == r {
+			found = true
```

Comment:
acha ruim ter um return true aqui e em baixo um return false? Já que vai dar break de qualquer forma no for?
Para ser sincero é indiferente pra mim, mas...